components/entity-access-rules/rest/src/main/java/org/phenotips/data/permissions/rest/internal/DefaultDomainObjectFactory.java

```diff
@@ -182,86 +204,239 @@ public PrincipalsRepresentation createPrincipalsRepresentation(PrimaryEntity ent
+        addEntity(entity, entityType, addedPrincipals, result, false);
 
-        Set<String> addedPrincipals = new LinkedHashSet<>();
+        return result;
+    }
 
+    /**
+     * Create PrincipalRepresentation of the entity's associated principals like owners, collaborators and/or family
+     * members, including all group and subgroup members.
+     *
+     * @param entity to whom we create principal representations
+     * @param entityType the type of the entity
+     * @param addedPrincipals cash of already added principals with access levels
+     * @param result REST representation for a list of principals that have access to the {@link PrimaryEntity}
+     * @param isFamily true if an entity is a family member
+     */
+    private void addEntity(PrimaryEntity entity, String entityType,
+        Map<String, Pair<PrincipalRepresentation, AccessLevel>> addedPrincipals, PrincipalsRepresentation result,
+        boolean isFamily)
+    {
+        EntityAccess entityAccess = this.manager.getEntityAccess(entity);
         Owner owner = entityAccess.getOwner();
-        Document doc = this.helper.getDocument(owner.getUser());
-        PrincipalRepresentation principal = this.createPrincipalRepresentation(owner.getUser(), owner.getType(),
-            "manage", "owner", "", doc.getURL());
-        result.withPrincipals(principal);
-        addedPrincipals.add(owner.getUsername());
+        String ownerName = owner.getUsername();
+
+        String role = "owner";
+        AccessLevel access = this.manageAccess;
+        PrincipalRepresentation principal = null;
+
+        // Adding entity owner. Two possible cases:
+        // 1. It is primary entity owner - in that case we do not need any checks
+        // 2. It is a family member owner
+
+        if (isFamily) {
+            if (addedPrincipals.containsKey(ownerName)) {
+                principal = addedPrincipals.get(ownerName).getLeft();
+                // - if this principal was already added as family collaborator with "view" access,
+                // then we update access level to "edit" (because owner of family member always
+                // has "edit" access rights to the family)
+                AccessLevel savedAccess = addedPrincipals.get(ownerName).getRight();
+                if (savedAccess.equals(this.viewAccess)) {
+                    checkAccessRight(addedPrincipals.get(ownerName).getLeft(), addedPrincipals.get(ownerName)
+                        .getRight(),
+                        this.editAccess, isFamily);
+                }
+            } else {
+                // - if this principal was never added and is Admin,
+                // then we save it save with access "manage" and role "admin".
+                DocumentReference ownerDocument = this.resolver.resolve(owner.getUser());
+                boolean isAdmin =
+                    this.rights.hasAccess(Right.ADMIN, ownerDocument, PermissionsConfiguration.PREFERENCES_DOCUMENT);
```

Comment:
@sdumitriu  By some reason that throws null pointer exception. Did I misuse hasAccess() somehow?